### PR TITLE
perf: optimize reading transactions in commit loop

### LINKED ISSUE #3057: Optimize IO scheduling for conflict resolution
When we commit a transaction, we load each of the transaction files from concurrent commits to check for conflicts. However, this is currently inefficient in two ways:

1. We do this completely synchronously, reading a transaction file and processing one at a time
    * Using a buffered read would be appropriate
    * We also might optimize the path to find the transaction file path. Right now, it requires reading a whole manifest file. It would be nice if we could get that path in fewer IOPs.
3. We don't seem to cache the transaction files, so each commit may have to re-open these transaction files.
    * A small TTL cache would be appropriate here 

The logic for this is at:
 https://github.com/lancedb/lance/blob/4812ac0d6dd0d15539e6f2690fb13a6491d2397f/rust/lance/src/io/commit.rs#L534-L565

You can see this clearly with the following repro:

```python
import lance
import pyarrow as pa
import numpy as np

from lance.tracing import trace_to_chrome
trace_to_chrome("trace.json", level="debug")

nrows = 10 * 1024
ntransactions = 100

data = pa.table({
    'id': range(nrows),
    'vectors': pa.FixedShapeTensorArray.from_numpy_ndarray(np.random.rand(nrows, 128))
})
data_size = data.get_total_buffer_size()
print(f"Data size: {data_size:,} bytes")

from concurrent.futures import ThreadPoolExecutor, as_completed

current_version = lance.dataset("test").version

with ThreadPoolExecutor(10) as executor:
    # Concurrently write fragments
    futures = [
        executor.submit(lance.fragment.write_fragments, data, "test")
        for _ in range(ntransactions)
    ]
    # Serially commit transactions
    for future in as_completed(futures):
        fragments = future.result()
        operation = lance.LanceOperation.Append(fragments)
        # read_version turns out to be critical for performance
        lance.LanceDataset.commit("test", operation, read_version=current_version)

```

### PR DESCRIPTION
**Pull request by wjones127**
* Cache transaction files
* Read transaction files with concurrency within commit loop

Closes #3057

### CODE REVIEW

File: rust/lance-core/src/cache.rs [hunk 1]
```diff
@@ -114,6 +114,7 @@ impl FileMetadataCache {
 
     pub fn size(&self) -> usize {
         if let Some(cache) = self.cache.as_ref() {
+            cache.sync();
```
Comment (westonpace) [inline rust/lance-core/src/cache.rs:117]: Is this mainly for testing purposes?  Or are there non-testing reasons we need this to be accurate?  I don't think it's a problem as we _shouldn't_ be calling `size` in a loop so seems fine, just curious.

File: rust/lance/src/dataset/write/commit.rs [hunk 2]
```diff
@@ -190,22 +190,23 @@ impl<'a> CommitBuilder<'a> {
...
-                    store_options: self.store_params.clone(),
-                    commit_handler: self.commit_handler.clone(),
-                    object_store_registry: self.object_store_registry.clone(),
-                    ..Default::default()
-                });
-
-                // If read_version is zero, then it might not have originally been
-                // passed. We can assume the latest version.
-                if transaction.read_version > 0 {
-                    builder = builder.with_version(transaction.read_version)
```
Comment (westonpace) [inline rust/lance/src/dataset/write/commit.rs:207]: Do we not need to pass in the read_version anymore?

File: rust/lance/src/io/commit.rs [hunk 3]
```diff
@@ -637,41 +685,43 @@ pub(crate) async fn commit_transaction(
...
+    } else {
+        dataset.checkout_latest().await?;
     }
+    let latest_version = dataset.manifest.version;
+    let other_transactions = futures::stream::iter((read_version + 1)..latest_version)
+        .map(|version| {
+            read_dataset_transaction_file(&dataset, version)
+                .map(move |res| res.map(|tx| (version, tx)))
+        })
+        .buffer_unordered(10)
```
Comment (westonpace) [inline rust/lance/src/io/commit.rs:702]: Why 10?  Can we use the store's I/O concurrency instead?

File: rust/lance/src/io/commit.rs [hunk 4]
```diff
@@ -743,18 +799,24 @@ pub(crate) async fn commit_transaction(
...
-                    check_transaction(transaction, version, &other_transaction)?;
-                }
+                dataset.checkout_latest().await?;
+                let latest_version = dataset.manifest.version;
+                futures::stream::iter(target_version..=latest_version)
+                    .map(|version| {
+                        read_dataset_transaction_file(&dataset, version)
+                            .map(move |res| res.map(|tx| (version, tx)))
+                    })
+                    .buffer_unordered(10)
```
Comment (westonpace) [inline rust/lance/src/io/commit.rs:809]: Same question as above.  Why 10?

File: rust/lance/src/io/commit.rs [hunk 5]
```diff
@@ -743,18 +799,24 @@ pub(crate) async fn commit_transaction(
...
+                    .and_then(|(version, other_transaction)| {
+                        let res = check_transaction(
+                            transaction,
+                            version,
+                            Some(other_transaction.as_ref()),
+                        );
+                        futures::future::ready(res)
+                    })
+                    .try_all(|_| futures::future::ready(true))
```
Comment (westonpace) [inline rust/lance/src/io/commit.rs:818]: Can you make the `and_then` a `try_for_each`?  Then you don't need the `try_all`?

File: rust/lance/src/utils/test.rs [hunk 6]
```diff
@@ -390,18 +399,47 @@ impl ObjectStore for IoTrackingStore {
...
+
+    fn put_part(&mut self, payload: PutPayload) -> UploadPart {
+        {
+            let mut stats = self.stats.lock().unwrap();
+            stats.write_iops += 1;
+            stats.write_bytes += payload.content_length() as u64;
+        }
+        self.target.put_part(payload)
+    }
+}
```
Comment (westonpace) [inline rust/lance/src/utils/test.rs:441]: Yay

**Review by westonpace — APPROVED** [review]
Just minor thoughts (though it looks like CI is failing)

File: rust/lance/src/dataset/write/commit.rs (hunk 2 above)
Comment (wjones127) [reply rust/lance/src/dataset/write/commit.rs:207]: I think, we only *need* to load the `read_version` if it's a detached commit. I've fixed this in the latest commits.

**Pull request merged**